diffrax/solver/ansr.py

```diff
@@ -0,0 +1,220 @@
+from dataclasses import dataclass
+from typing import Tuple
+
+import diffrax
+import jax
+import jax.lax as lax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+import numpy as np
+from equinox.internal import ω
+
+from ..custom_types import Bool, DenseInfo, LevyVal, PyTree, Scalar
+from ..local_interpolation import LocalLinearInterpolation
+from ..solution import RESULTS
+from ..term import AbstractTerm, ControlTerm, MultiTerm, ODETerm, WrapTerm
+from .base import AbstractItoSolver
+
+
+_ErrorEstimate = None
+_SolverState = None
+
+
+def check_diffusion_has_stla(diffusion):
+    # ensure that STLA will be computed
+    if isinstance(diffusion, WrapTerm):
+        if isinstance(diffusion.term, ControlTerm):
+            control = diffusion.term.control
+        else:
+            raise TypeError("diffusion should be a ControlTerm")
+    else:
+        control = diffusion.control
+    if not isinstance(control, diffrax.VirtualBrownianTree):
+        raise TypeError(
+            "A VirtualBrownianTree should be used " "to control the diffusion term."
+        )
+    elif not control.compute_stla:
+        raise ValueError(
+            "The VirtualBrownianTree controlling the diffusion"
+            ' should be initialised with "compute_stla=True"'
+        )
+
+
+@dataclass(frozen=True)
+class StochasticButcherTableau:
+    """A Butcher Tableau for Additive-noise SRK methods."""
+
+    # Only supports explicit SRK so far
+    c: np.ndarray
+    b: np.ndarray
+    a: list[np.ndarray]
+
+    # coefficients for W and H (of shape (len(c)+1,)
+    cw: np.ndarray
+    ch: np.ndarray
+    cw_last: Scalar
+    ch_last: Scalar
+
+    def __post_init__(self):
+        assert self.c.ndim == 1
+        for a_i in self.a:
+            assert a_i.ndim == 1
+        assert self.b.ndim == 1
+        assert self.c.shape[0] == len(self.a)
+        assert all(i + 1 == a_i.shape[0] for i, a_i in enumerate(self.a))
+        assert self.c.shape[0] + 1 == self.b.shape[0]
+        assert self.cw.shape[0] == self.b.shape[0]
+        assert self.ch.shape[0] == self.b.shape[0]
+        for i, (a_i, c_i) in enumerate(zip(self.a, self.c)):
+            assert np.allclose(sum(a_i), c_i)
+        assert np.allclose(sum(self.b), 1.0)
+
+        # TODO: add checks for whether the method is FSAL
+
+
+class ANSR(AbstractItoSolver):
+    r"""Additive-Noise Stochastic Runge-Kutta method.
```

Comment:
By the way, what's the reason for restricting to only additive noise? I think the SRK schemes themselves don't change that much.

Reply:
I Intend to add that eventually, but for now this will have to do. The way Rößler defines his tableaus is incredibly convoluted, and from what James told me, they might be a bit of a dead end, so it isn't the highest priority on my list right now.

Reply:
In fact I am in the proccess of generalising this, expect an update of that together with the space-time-time Levy area sometime next month.